feat: support regex in matchers

andrius-girdzius opened the pull request:
## Description

**WIP**

<!--
Thank you for contributing!

Step 1: Before submitting a pull request that introduces a new functionality or fixes a bug,
please open an issue where we could discuss the suggestion, problem - and potential ways to fix.
-->

- This pull request addresses the issue described here: #3366 

<!--
Step 2: Provide an overview of how your fix / enhancement works.
If possible, provide screenshots of the before and after states (even for simple command line options - show the terminal).
-->

In this pull request, I have added support for Regex text/label/testID matching in Detox.

<!--
Step 3: Please review the checklist below.
-->

---


> _For features/enhancements:_
 - [x] I have added/updated the relevant references in the [documentation](https://github.com/wix/Detox/tree/master/docs) files.

> _For API changes:_
 - [x] I have made the necessary changes in the [types index](https://github.com/wix/Detox/blob/master/detox/index.d.ts) file.

asafkorem commented on detox/android/detox/src/full/java/com/wix/detox/espresso/matcher/ViewMatchers.kt:
```diff
@@ -54,6 +64,42 @@ fun toHaveSliderPosition(expectedValuePct: Double, tolerance: Double): Matcher<V
         }
     }
 
+fun withRegexText(jsRegex: String): Matcher<View> =
+    object: BoundedMatcher<View, TextView>(TextView::class.java) {
+        override fun describeTo(description: Description) {
+            description.appendText("withRegexText($jsRegex)")
+        }
+
+        override fun matchesSafely(view: TextView): Boolean {
+            val text = view.text?.toString()
+            return text?.isMatch(jsRegex) ?: false
+        }
+    }
+
+fun withRegexTagValue(jsRegex: String): Matcher<View> =
+    object: BoundedMatcher<View, View>(View::class.java) {
+        override fun describeTo(description: Description) {
+            description.appendText("withRegexTagValue($jsRegex)")
+        }
+
+        override fun matchesSafely(view: View): Boolean {
+            val tag = view.tag?.toString()
+            return tag?.isMatch(jsRegex) ?: false
+        }
+    }
+
+fun withRegexContentDescription(jsRegex: String): Matcher<View> =
+    object: BoundedMatcher<View, View>(View::class.java) {
+        override fun describeTo(description: Description) {
+            description.appendText("withRegexContentDescription($jsRegex)")
+        }
+
+        override fun matchesSafely(view: View): Boolean {
+            val contentDescription = view.contentDescription?.toString()
+            return contentDescription?.isMatch(jsRegex) ?: false
+        }
+    }
```
You can avoid this duplicated structure by adding a factory method for those regex-matchers:

```
private fun createRegexMatcher(matcherDescription: String, regex: String, extractStringFromView: (View) -> String?): Matcher<View> =
    object: BoundedMatcher<View, View>(View::class.java) {
        override fun describeTo(description: Description) {
            description.appendText("$matcherDescription($regex)")
        }

        override fun matchesSafely(view: View): Boolean {
            val extractedString = extractStringFromView(view)
            return extractedString?.isMatch(regex) ?: false
        }
    }
```

It might even be worth extracting it to another file 🤔

asafkorem replied:
Usage example:

```kotlin
private fun withRegexText(jsRegex: String): Matcher<View> =
    createRegexMatcher("withRegexText", jsRegex) { view -> (view as? TextView)?.text?.toString() }
```

asafkorem replied:
Also, those methods should be private since you're using them in this file context only

asafkorem replied:
Another small note, sorry if it seems nitpicky, but personally, I prefer ordering methods according to their "caller-callee" relationship, where a method follows its caller when possible (if the method has not been declared before). This approach improves readability and mirrors the logical flow of execution in the code

For example, in this order, `withRegexText` will follow `withText`

asafkorem commented on detox/android/detox/src/full/java/com/wix/detox/espresso/matcher/ViewMatchers.kt:
```diff
@@ -54,6 +64,42 @@ fun toHaveSliderPosition(expectedValuePct: Double, tolerance: Double): Matcher<V
+
+        override fun matchesSafely(view: TextView): Boolean {
+            val text = view.text?.toString()
+            return text?.isMatch(jsRegex) ?: false
```
Have you considered replacing the custom `withText` method with this one? Instead of deciding between different matchers (`withRegexText` and Espresso's `withText`) prior to their invocation, you could simply use an if-else in the return value, like this: return `isRegex ? (value?.isMatchJSRegex(expected) ?: false) : value == expected`. Same Q for the other regex-matchers

asafkorem replied:
Honestly I'm not sure if it will be better, WDYT?

andrius-girdzius replied:
Yeah I believe it would be nicer code-wise to eliminate use of Espresso's `withText`.

My main concern was changing too much - I only briefly looked at the Espresso's implementation so I decided to simply add on regex matching and use their implementation elsewhere, just in case my own implementation would break existing Detox consumers. 

But I think I can open the source code again and have a deeper look to eliminate use of Espresso here, if you think that's okay 👍

asafkorem replied:
Generally, that's a great thinking. It makes sense not to disrupt existing usage. 
A thorough implementation, however, poses no risk if done correctly

asafkorem replied:
The simpler their implementation, the safer it is to "copy-paste" it and change it according to our needs..
If it adds too much code to our implementation it's suspicious, because it looks like a simple `value == expected` check, then we might miss something in our regex implementation 😅 

I prefer the simpler solution for readability and cleanliness, as you see fit

asafkorem commented on detox/android/detox/src/full/java/com/wix/detox/espresso/matcher/RegexMatcher.kt:
```diff
@@ -0,0 +1,42 @@
+// Other flags (e.g. g,u,s) are not supported as they do not have equivalents in Kotlin.
+//
+// - See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
+fun String.isMatch(jsRegex: String): Boolean {
```
I think calling it `matchesJSRegex` would be more self-explanatory. `isMatch` is too broad

asafkorem replied:
I know I know, that's what I called it in iOS; I take the responsibility. The same comment applies 😅

asafkorem commented on detox/android/detox/src/full/java/com/wix/detox/espresso/matcher/RegexMatcher.kt:
```diff
@@ -0,0 +1,42 @@
+    return Regex(pattern, options).matches(this)
+}
+
+fun getRegexFlags(jsRegex: String): CharSequence {
```
add the `private` modifier where possible

asafkorem commented on detox/src/android/AndroidExpect.test.js:
```diff
@@ -43,6 +43,9 @@ describe('AndroidExpect', () => {
       await e.expect(e.element(e.by.accessibilityLabel('test'))).toHaveText('text');
       await e.expect(e.element(e.by.accessibilityLabel('test'))).toNotHaveText('text');
       await e.expect(e.element(e.by.accessibilityLabel('test'))).not.toHaveText('text');
+      await e.expect(e.element(e.by.accessibilityLabel('test'))).toHaveText(/text/);
+      await e.expect(e.element(e.by.accessibilityLabel('test'))).toNotHaveText(/text/);
+      await e.expect(e.element(e.by.accessibilityLabel('test'))).not.toHaveText(/text/);
```
add the same for label and test-id matchers

asafkorem commented on detox/src/android/espressoapi/DetoxMatcher.js:
```diff
@@ -14,39 +14,51 @@ function sanitize_matcher(matcher) {
 class DetoxMatcher {
-  static matcherForText(text) {
+  static matcherForText(text, isRegex) {
     if (typeof text !== "string") throw new Error("text should be a string, but got " + (text + (" (" + (typeof text + ")"))));
+    if (typeof isRegex !== "boolean") throw new Error("isRegex should be a boolean, but got " + (isRegex + (" (" + (typeof isRegex + ")"))));
```
Consider extracting these assertions into a separate method. Doing so would:
- Minimize code duplication
- Adhere to the Single Responsibility Principle (SRP)

andrius-girdzius replied:
this file is autogenerated based on params on `DetoxMatcher.java`. Did you mean I should edit the generation code?

asafkorem replied:
ahh, right. never mind then 😆

asafkorem reviewed (changes requested):
You've done some good work here, I'm really excited to see this! 
Kindly let me know when it's ready for another review, post exiting draft mode

asafkorem commented:
Upon quick review, I must say that this is a masterpiece. Great work! 🙌🏼 👑 
I might have some minor comments but it's quite there, I'll do another round later today / tomorrow

andrius-girdzius commented:
> Upon quick review, I must say that this is a masterpiece. Great work! 🙌🏼 👑 I might have some minor comments but it's quite there, I'll do another round later today / tomorrow

@asafkorem Sounds great! Looking forward to the comments, fixing and merging! 🥳

asafkorem commented on docs/api/matchers.md:
````diff
@@ -42,6 +42,20 @@ Match elements with the specified text.
 element(by.text('Tap Me'));
 ```
 
+You can also use regex with support of the following JS flags:
+
+| Flag | Name          | Modification                                                                                                                                       |
+|------|---------------|----------------------------------------------------------------------------------------------------------------------------------------------------|
+| `i`  | Ignore Casing | Makes the expression search case-insensitively.                                                                                                    |
+| `s`  | Dot All       | Makes the wild character . match newlines as well.                                                                                                 |
+| `m`  | Multiline     | Makes the boundary characters ^ and $ match the beginning and ending of every single line instead of the beginning and ending of the whole string. |
+
+**Note:** JS flags like `g` (global) and `y` (sticky) that are not supported, as well as `u` (unicode) which is always implied, are ignored when parsing input.
+
+```js
+element(by.text(/Tap [A-Za-z]+/i));
````
I love it. Minor suggestion for improvement:

```suggestion
You can also utilize regex (Regular Expressions) alongside certain flags. Here's a table with the supported flags:

| Flag | Name          | Modification                                                                                                                                       |
|------|---------------|----------------------------------------------------------------------------------------------------------------------------------------------------|
| `i`  | Ignore Casing | Makes the expression search case-insensitively.                                                                                                    |
| `s`  | Dot All       | Makes the wild character `.` match newlines as well.                                                                                                 |
| `m`  | Multiline     | Makes the boundary characters (`^` and `$`) match the beginning and ending of every single line instead of the beginning and ending of the whole string. |

**Note:** JS flags like `g` (global) and `y` (sticky) that are not supported, as well as `u` (unicode) which is always implied, are ignored when parsing input.

The following sample code snippet matches text starting with "Tap" followed by any number of alphabetic characters, case-insensitively:

```js
element(by.text(/Tap [A-Za-z]+/i));
```

asafkorem reviewed (changes requested):
Again, great work!! I'm super excited with this! Just a few small tweaks and we can ship it :shipit:

asafkorem commented on detox/android/detox/src/full/java/com/wix/detox/espresso/matcher/RegexMatcher.kt:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    companion object {
+        fun <T> matchesRegex(jsRegex: String): RegexMatcher<T> {
```
However not sure if we need that

andrius-girdzius replied:
ah, that's true, it's not really needed here 💡

andrius-girdzius commented:
> Again, great work!! I'm super excited with this! Just a few small tweaks and we can ship it :shipit:

I've added the fixes, also modified `matchers.md` slightly to have a separate section on regex matching (since it is relevant for multiple matchers) and added links to it from the supported methods. 🥼

asafkorem commented:
> > Again, great work!! I'm super excited with this! Just a few small tweaks and we can ship it :shipit:
> 
> I've added the fixes, also modified `matchers.md` slightly to have a separate section on regex matching (since it is relevant for multiple matchers) and added links to it from the supported methods. 🥼

GG!

noomorph commented on detox/src/ios/expectTwo.js:
```diff
@@ -10,6 +10,7 @@ const { assertEnum, assertNormalized } = require('../utils/assertArgument');
 const { removeMilliseconds } = require('../utils/dateUtils');
 const { actionDescription, expectDescription } = require('../utils/invocationTraceDescriptions');
 const log = require('../utils/logger').child({ cat: 'ws-client, ws' });
+const isRegExp = require('../utils/regexp');
```
For consistency, I would suggest:

1) to rename the file to `utils/isRegExp.js`
2) use named exports, e.g.: `const { isRegExp } = require(...)`

noomorph commented on detox/src/utils/regexp.js:
```diff
@@ -0,0 +1,5 @@
+function isRegExp(obj) {
+  return Object.prototype.toString.call(obj) === '[object RegExp]';
```
I assume this is because of the different sandboxes (Jest), right?

andrius-girdzius replied:
Yes, simply checking `instanceof` fails in this case, so I had to use a workaround

noomorph commented on docs/api/matchers.md:
```diff
@@ -117,3 +135,21 @@ On iOS, matched elements are sorted by their x and y axes.
+| `s`  | Dot All       | Makes the wild character `.` match newlines as well.                                                                                                     |
+| `m`  | Multiline     | Makes the boundary characters (`^` and `$`) match the beginning and ending of every single line instead of the beginning and ending of the whole string. |
+
+**Note:** JS flags like `g` (global) and `y` (sticky) that are not supported, as well as `u` (unicode) which is always implied, are ignored when parsing input.
```
I'd suggest to create an admonition and warn about possible issues with look-behind expressions (iOS/Swift doesn't support them where as Kotlin supports), e.g.:

![image](https://github.com/wix/Detox/assets/1962469/11c7ea36-f5ad-4cf3-b313-0edd55d63577)

asafkorem replied:
@noomorph sounds good. Great catch with the "look behind" assertions! IMO we should provide this information inline.

```suggestion
:::caution Note

Regular expression flags such as `g` (global) and `y` (sticky) that are not supported, as well as `u` (unicode) which is always implied, are ignored when parsing input. 

Pay attention that as of writing this note, Android supports lookbehind assertions in its regular expression implementation, while iOS does not. It's advisable to check the platform-specific documentation for limitations.

:::
```

asafkorem replied:
Due to the fact that this is only a button tap (+ to avoid the bikeshedding), I will allow myself to commit this change 🙂

asafkorem commented:
Released with v20.11.0 🎉 
Thanks again @andrius-girdzius! 

![7rioa1](https://github.com/wix/Detox/assets/55082339/3cef8192-1dbe-48d2-b505-95a90815e26d)
